lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -384,15 +385,29 @@ matchWitnessDifferentiableAttr(DeclContext *dc, ValueDecl *req,
     }
     if (!foundExactConfig) {
       bool success = false;
-      if (supersetConfig) {
-        // If the witness has a "superset" derivative configuration, create an
-        // implicit `@differentiable` attribute with the exact requirement
-        // `@differentiable` attribute parameter indices.
+      // If no exact witness derivative configuration was found, check
+      // conditions for creating an implicit witness `@differentiable` attribute
+      // with the exact derivative configuration:
+      // - If the witness has a "superset" derivative configuration.
+      // - If the witness is less than public and is declared in the same file
+      //   as the conformance.
+      //   - `@differentiable` attributes are really only significant for public
+      //     declarations: it improves usability to not require explicit
+      //     `@differentiable` attributes for less-visible declarations.
+      bool createImplicitWitnessAttribute =
+          supersetConfig ||
+          (witness->getFormalAccess() < AccessLevel::Public &&
+           dc->getModuleScopeContext() ==
```

Comment:
Note: we could add a more specific diagnostic for "witness has less-than-public visibility but is declared in a different file than the conformance".

We need a good diagnostic message though. Suggestions welcome:
> note: candidate is missing attribute '@differentiable'; explicit '@differentiable' attribute is required for non-public declaration 'foo' because it is in a different file than the 'ConformingType: Protocol' conformance

Reply:
Feels like you've squashed two different errors into one. Instead, could you use one error for _non-public declarations in other files_, and a different error for _public declarations_?

Reply:
Done in https://github.com/apple/swift/pull/29771/commits/e22a4e8914c233a3b83e718bd093a26a7d4dbd74.

Here's the dedicated diagnostic for "non-public declarations in other files":
```console
main.swift:28:1: error: type 'ConformingStruct' does not conform to protocol 'Protocol'
extension ConformingStruct: Protocol {}
^
Inputs/other_file.swift:33:8: note: non-public instance method 'internalMethod1' must have explicit '@differentiable' attribute to satisfy requirement instance method 'internalMethod1' (in protocol 'Protocol') because it is declared in a different file than the conformance of 'ConformingStruct' to 'Protocol'
  func internalMethod1(_ x: Float) -> Float {
       ^
  @differentiable
```
> non-public instance method 'internalMethod1' must have explicit '@differentiable' attribute to satisfy requirement instance method 'internalMethod1' (in protocol 'Protocol') because it is declared in a different file than the conformance of 'ConformingStruct' to 'Protocol'

It's pretty long but accurate. I copied the template from other witness/requirement mismatch diagnostics.